src/parameterFiles.ts

```diff
@@ -0,0 +1,507 @@
+    await setMappedParamFileForTemplate(templateUri, result.data?.uri);
+  }
+}
+
+export async function openParameterFile(actionContext: IActionContext, sourceUri: Uri | undefined): Promise<void> {
+  if (sourceUri) {
+    let paramFile: Uri | undefined = findMappedParamFileForTemplate(sourceUri);
+    if (!paramFile) {
+      throw new Error(`There is no parameter file currently associated with template file "${sourceUri.fsPath}"`);
+    }
+
+    let doc: TextDocument = await workspace.openTextDocument(paramFile);
+    await window.showTextDocument(doc, ViewColumn.Beside);
+  }
+}
+
+/**
+ * If the params file is inside the workspace folder, use the path relative to its template file. Otherwise, return the
+ * absolute path to the params file. This is intended to make the path most logical to the user.
+ */
+export function getFriendlyPathToParamFile(templateUri: Uri, paramFileUri: Uri): string {
+  const workspaceFolder = workspace.getWorkspaceFolder(paramFileUri);
+  if (workspaceFolder) {
+    return path.relative(path.dirname(templateUri.fsPath), paramFileUri.fsPath);
+  } else {
+    return paramFileUri.fsPath;
+  }
+}
+
+interface QuickPickList {
+  items: IAzureQuickPickItem<IPossibleParamFile | undefined>[];
+  currentParamFile: IPossibleParamFile | undefined;
+  none: IAzureQuickPickItem<IPossibleParamFile | undefined>;
+  browse: IAzureQuickPickItem<IPossibleParamFile | undefined>;
+  openCurrent: IAzureQuickPickItem<IPossibleParamFile | undefined>;
+}
+
+async function createParamFileQuickPickList(templateUri: Uri): Promise<QuickPickList> {
+  // Find likely parameter file matches
+  let suggestions: IPossibleParamFile[] = await findSuggestedParameterFiles(templateUri);
+
+  // Find the current in that list
+  const currentParamUri: Uri | undefined = findMappedParamFileForTemplate(templateUri);
+  const currentParamPathNormalized: string | undefined = currentParamUri ? normalizePath(currentParamUri) : undefined;
+  let currentParamFile: IPossibleParamFile | undefined = suggestions.find(pf => normalizePath(pf.uri) === currentParamPathNormalized);
+  if (currentParamUri && !currentParamFile) {
+    // There is a current parameter file, but it wasn't among the list we came up with.  We must add it to the list.
+    currentParamFile = { isCloseNameMatch: false, uri: currentParamUri, friendlyPath: getFriendlyPathToParamFile(templateUri, currentParamUri) };
+    suggestions = suggestions.concat(currentParamFile);
+  }
+
+  // Create initial pick list and sort it
+  let pickItems: IAzureQuickPickItem<IPossibleParamFile | undefined>[] = suggestions.map(paramFile => createQuickPickItem(paramFile, currentParamFile, templateUri));
+  sortQuickPickList(pickItems);
+
+  // Move the current item (if any) to the top
+  const currentItem = pickItems.find(pi => pi.data === currentParamFile);
+  if (currentItem) {
+    // tslint:disable-next-line: no-any
+    pickItems = [currentItem].concat(pickItems.filter(ppf => ppf !== currentItem));
```

Comment:
>pickItems [](start = 4, length = 9)

Nit: Can't we just use array indexes to swap instead of rebuilding the array?

Reply:
Yeah, but this is easy to understand, less likely to get wrong, and performance not an issue.